IDR/notebook/scripts/nbconverted/2.collect_statistics.py

```diff
@@ -0,0 +1,354 @@
+
+
+# In[16]:
+
+
+def walk(path):
+    """Collects paths for files in path parameter
+
+    Parameters
+    ----------
+    path: str or pathlib.Path() object
+            Path to metadata folder containing IDR study directories
+
+    Returns
+    -------
+    PosixPath object
+    """
+    for subdir in pathlib.Path(path).iterdir():
+        if subdir.is_dir():
+            yield from walk(subdir)
+            continue
+        yield subdir.resolve()
+
+
+# In[17]:
+
+
+# Define stats
+def category_frequencies(attribute_elements):
+    """Calculates absolute and relative frequencies for unique elements of an image attribute
+
+    Parameters
+    ----------
+    attribute_elements: dict
+        Unique elements of an image attribute as keys and instances of each as values
+
+    Returns
+    -------
+    rel_freq_dict: dict
+        Relative frequencies per unique image attribute element
+    abs_freq-list: list
+        Instance counts of unique image attribute elements
+    """
+    total_instances = sum(attribute_elements.values())
+    rel_freq_dict = dict()
+    abs_freq_list = list()
+    for image_attribute in attribute_elements.keys():
+        abs_freq_list.append(attribute_elements[image_attribute])
+        rel_freq_dict[image_attribute] = (
+            attribute_elements[image_attribute] / total_instances
+        )
+    return rel_freq_dict, abs_freq_list
+
+
+def h_index(p):
+    """Calculates the Shannon Index of a set of unique attribute instances.
+
+    Parameters
+    ----------
+    p: dict
+        Dictionary of relative frequencies for each unique element in an attribute column.
+
+    Returns
+    -------
+    h: float
+        Shannon Index value
+    results: list
+        List of each -p_iln(p_i) value to use for Normalized Median Evenness statistic.
+    """
+    results = list()
+    for entry in p.values():
+        results.append(entry * ln(entry))
+
+    results = np.array(results)
+    h = -(sum(results))
+    return h, results
+
+
+def pielou(h, s):
+    """Calculates the ratio of the observed Shannon Index to the maximum possible Shannon Index value
+
+    Parameters
+    ----------
+    h: float
+        Observed Shannon Index calculated by h-index()
+    s: int
+        Observed richness within an image attribute (count of unique entries)
+
+    Returns
+    -------
+    j: float
+        Pielou's evenness (H_obs / H-max)
+    """
+    j = h / ln(s)
+    return j
+
+
+def norm_median_evenness(h_list):
+    """Calculates Normalized Median Evenness (NME) of Shannon Index summation elements (-p*ln(p))
+
+    Parameters
+    ----------
+    h_list: list
+        Individual -p*ln(p) values of the Shannon Index summation
+
+    Returns
+    -------
+    nme: float
+        Ratio of median and max -p*ln(p) values
+    """
+    # Multiply each value in h_list by -1
+    h_values = np.array([-1.0 * h_value for h_value in h_list])
+
+    # Calculate NME
+    nme = ndimage.median(h_values) / h_values.max()
```

Comment:
Looks like there are many ways to calculate evenness: http://scikit-bio.org/docs/0.4.2/generated/skbio.diversity.alpha.html?highlight=evenness